doc/design/cluster_train/trainer.md

```diff
@@ -0,0 +1,82 @@
+# Design Doc: Trainer Communication Library
+
+For an overview of trainer's role, please refer to [distributed training design doc](README.md). In this design doc, we will discuss the trainer's communication library, which will manage communication with parameter servers and the [master server](master_server.md). The library will be implemented in [Go](https://golang.org/) and made available as a static or dynamic library with a C header file.
+
+## Go Interface
+
+The Go interface is the basic abstraction of communications with the master server and parameter servers. We will add another layer on top (add retry logic, polish interface with C idiom) before exposing the library with a [C interface](#c-interface).
+
+```go
```

Comment:
The client library is not just a simple wrapper on top of RPC. For example `ParamInitChan` will use an etcd lock to figure out who will init the parameters.

Indeed, the API here highly depend on the parameter server RPC API. So this PR and #1964 depends on each other. This in the API in my mind, I will be discussing with @dzhwinter and everyone to form the same API, borrowing ideas from both PRs.